### PR TITLE
CDAP-13774 add remote hadoop provisioner

### PR DESCRIPTION
**Pull request by albertshau**
This provisioner just connects to an existing cluster instead of
actually creating and deleting a cluster. It needs to be configured
with the host to connect to, as well as the ssh user and keys
required to connect.

As part of this work, moved the standard node properties that CDAP
expects into the spi instead of in the dataproc module. Also fixing
a typo to change the dataproc name from gce-dataproc to gcp-dataproc.

### CODE REVIEW

File: cdap-runtime-ext-dataproc/src/main/java/co/cask/cdap/runtime/spi/provisioner/dataproc/DataProcProvisioner.java [hunk 1]
```diff
@@ -44,7 +45,7 @@
 
   private static final Logger LOG = LoggerFactory.getLogger(DataProcProvisioner.class);
   private static final ProvisionerSpecification SPEC = new ProvisionerSpecification(
-    "gce-dataproc", "Google Cloud Dataproc Provisioner",
```
Comment (nitinmotgi) [inline cdap-runtime-ext-dataproc/src/main/java/co/cask/cdap/runtime/spi/provisioner/dataproc/DataProcProvisioner.java:47]: Remove provisioner. "Google Cloud Dataproc"

File: cdap-runtime-ext-dataproc/src/main/java/co/cask/cdap/runtime/spi/provisioner/dataproc/DataProcProvisioner.java [hunk 2]
```diff
@@ -44,7 +45,7 @@
...
+    "gcp-dataproc", "Google Cloud Dataproc Provisioner",
```
Comment (bdmogal) [inline cdap-runtime-ext-dataproc/src/main/java/co/cask/cdap/runtime/spi/provisioner/dataproc/DataProcProvisioner.java]: Can you also remove the word "Provisioner" from here?

Comment (albertshau) [reply cdap-runtime-ext-dataproc/src/main/java/co/cask/cdap/runtime/spi/provisioner/dataproc/DataProcProvisioner.java]: there's a separate jira for that, I probably should have left this out of this PR.

File: cdap-runtime-ext-dataproc/src/main/java/co/cask/cdap/runtime/spi/provisioner/dataproc/DataProcProvisioner.java (hunk 1 above)
Comment (albertshau) [reply cdap-runtime-ext-dataproc/src/main/java/co/cask/cdap/runtime/spi/provisioner/dataproc/DataProcProvisioner.java:47]: there's a separate jira for that

File: cdap-runtime-ext-remote-hadoop/src/main/resources/remote-hadoop.json [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+          "name": "user",
+          "required": true,
+          "description": "The user to SSH to the cluster as.",
+          "widget-attributes": {
+            "placeholder": "Specify your User ID"
+          }
+        },
+        {
+          "widget-type": "textarea",
+          "label": "SSH Public Key",
```
Comment (anwar6953) [inline cdap-runtime-ext-remote-hadoop/src/main/resources/remote-hadoop.json]: why does the user need to configure the public key?

Comment (anwar6953) [reply cdap-runtime-ext-remote-hadoop/src/main/resources/remote-hadoop.json]: In order to ssh into a machine, you don't need the public key, right?

**Comment by albertshau** [comment]
https://builds.cask.co/browse/CDAP-RUT1551

File: cdap-runtime-ext-remote-hadoop/src/main/resources/remote-hadoop.json (hunk 3 above)
Comment (albertshau) [reply cdap-runtime-ext-remote-hadoop/src/main/resources/remote-hadoop.json]: right, you don't. I thought SSHContext.createSSHSession() was using it somehow but I see now it just uses the 'user' field in SSHPublicKey. Will remove it

File: cdap-runtime-spi/src/main/java/co/cask/cdap/runtime/spi/Constants.java [hunk 4]
```diff
@@ -12,9 +12,10 @@
  * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
  * License for the specific language governing permissions and limitations under
  * the License.
+ *
  */
 
-package co.cask.cdap.runtime.spi.provisioner.dataproc;
+package co.cask.cdap.runtime.spi;
```
Comment (anwar6953) [inline cdap-runtime-spi/src/main/java/co/cask/cdap/runtime/spi/Constants.java:18]: +1
Thanks for moving this.

File: cdap-runtime-ext-remote-hadoop/src/main/resources/remote-hadoop.json (hunk 3 above)
Comment (anwar6953) [reply cdap-runtime-ext-remote-hadoop/src/main/resources/remote-hadoop.json]: Actually, it does seem to be used here:
https://github.com/caskdata/cdap/blob/release/5.0/cdap-app-fabric/src/main/java/co/cask/cdap/internal/provision/DefaultSSHContext.java#L88

Though, technically, for the remote hadoop provisioner, the value is written to a file there, but not actually ever used.

So, we can probably use an empty string, but not null.

Or we should mark the public key `@Nullable`, and then change it to not write to file if its null.

**Review by anwar6953 — APPROVED** [review]

**Pull request merged**